rechunker/executors/pywren.py

```diff
@@ -0,0 +1,70 @@
+from functools import partial
+
+from typing import Callable, Iterable, Tuple
+
+from rechunker.executors.util import chunk_keys, split_into_direct_copies
+from rechunker.types import CopySpec, Executor, ReadableArray, WriteableArray
+
+import pywren_ibm_cloud as pywren
+from pywren_ibm_cloud.executor import FunctionExecutor
+
+# PywrenExecutor represents delayed execution tasks as functions that require
+# a FunctionExecutor.
+Task = Callable[[FunctionExecutor], None]
+
+
+class PywrenExecutor(Executor[Task]):
+    """An execution engine based on Pywren.
+
+    Supports zarr arrays as inputs. Outputs must be zarr arrays.
+
+    Any Pywren FunctionExecutor can be passed to the constructor. By default
+    a Pywren `local_executor` will be used
+
+    Execution plans for PywrenExecutor are functions that accept no arguments.
+    """
+
+    def __init__(self, pywren_function_executor: FunctionExecutor = None):
+        self.pywren_function_executor = pywren_function_executor
+
+    def prepare_plan(self, specs: Iterable[CopySpec]) -> Task:
+        tasks = []
+        for spec in specs:
+            for direct_spec in split_into_direct_copies(spec):
+                tasks.append(partial(_direct_array_copy, *direct_spec))
+        return partial(_execute_all, tasks)
+
+    def execute_plan(self, plan: Task):
+        if self.pywren_function_executor is None:
+            # No Pywren function executor specified, so use a local one, and shutdown after use
+            with pywren.local_executor(
+                # Minimal config needed to avoid Pywren error if ~/.pywren_config is missing
+                config={"pywren": {"storage_bucket": "unused"}}
+            ) as pywren_function_executor:
+                plan(pywren_function_executor)
+        else:
+            plan(self.pywren_function_executor)
+
+
+def _direct_array_copy(
+    source: ReadableArray,
+    target: WriteableArray,
+    chunks: Tuple[int, ...],
+    pywren_function_executor: FunctionExecutor,
+) -> None:
+    """Direct copy between arrays using Pywren for parallelism"""
+    iterdata = [(source, target, key) for key in chunk_keys(source.shape, chunks)]
+
+    def direct_copy(iterdata):
+        source, target, key = iterdata
+        target[key] = source[key]
+
+    pywren_function_executor.map(direct_copy, iterdata)
+    pywren_function_executor.get_result()
```

Comment:
This looks like it makes the computation blocking?

If so, my main concern is what happens if it gets run on a group containing multiple Zarr arrays. The different arrays can be safely copied in parallel, which could potentially be much more efficient. Only the two stages of copying a single array need to happen separately.

Reply:
I didn't really think about groups (as you and @TomAugspurger rightly pointed out!). I've updated the PR to split the computation into serial and parallel parts.

Reply:
The tests are failing intermittently, which I suspect may be due to the concurrent executor I introduced in the previous commit, and how it interacts with Pywren's threading (although I haven't been able to confirm that). Rather than fix that now, I wonder if it would be reasonable to merge this without the parallel support for Zarr groups? (I'd like to see if there is a way to avoid using a concurrent executor, and just use the Pywren futures that it returns, but that will take a bit more work.)

I've updated the code to add a note about this, and also to leave the function with the name `_execute_in_series` to make the limitation clearer.

Reply:
That works for me!